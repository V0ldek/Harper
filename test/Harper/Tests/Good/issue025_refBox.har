ref RefBox a = {
    data = {
        a :: a;
    }

    ctor :: a -> impure -> RefBox a;
    ctor a = {
        this.data.a := a;
    };

    get :: impure -> a;
    get = this.data.a;

    set :: a -> impure -> ();
    set a = {
        this.data.a := a;
    };
};

f :: RefBox Integer -> impure -> ();
f b = set 0 b ();

main :: sideeffect -> Integer;
main = {
    (b :: RefBox Integer) = RefBox.ctor 42 ();

    eval printLn (get b) ();     // 42

    (set17 :: impure -> ()) = set 17 b;

    eval printLn (get b) ();     // 42

    eval set17 ();

    eval printLn (get b) ();     // 17

    (b' :: RefBox Integer) = b;

    eval printLn (get b') ();    // 17

    eval set 42 b ();

    eval printLn (get b') ();    // 42
    eval printLn (get b) ();     // 42

    eval f b ();

    eval printLn (get b') ();    // 0
    eval printLn (get b) ();     // 0

    return get b ();
};