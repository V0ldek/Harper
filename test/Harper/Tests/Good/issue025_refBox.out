42
42
17
17
42
42
0
0

Execution ended with value: 0